### PR TITLE
Fix string-based route types not saving (Fixes: #5743)

### PR DESCRIPTION
**Pull request by tsimonq2**
Fixes: #5743

### CODE REVIEW

File: sql/database.sql
Comment (sjpadgett) [inline sql/database.sql:7719]: This change worries me! There's a reason this is an integer that only the original dev knows however we can't know what ramifications this change will have on other aspects of the module e.g. reporting, laminas module or validation.

File: sql/7_0_0-to-7_0_1_upgrade.sql
Comment (sjpadgett) [inline sql/7_0_0-to-7_0_1_upgrade.sql:198]: Fix this so it is consistent with the rest of the list. 'bymouth' to 25. Save as integer not string.
We have other lists with similar id pattern i.e Drug Interval.

Simply making this change should fix the original issue. At least it did for me.

File: library/classes/Prescription.class.php
Comment (sjpadgett) [inline library/classes/Prescription.class.php:379]: This seems purposeful where someone though it important enough to do the numeric check

File: sql/database.sql
Comment (stephenwaite) [reply sql/database.sql:7719]: another way to approach it could be to grab the list_id instead in [the template](https://github.com/openemr/openemr/blob/master/templates/prescription/general_edit.html#L231)

File: sql/7_0_0-to-7_0_1_upgrade.sql
Comment (stephenwaite) [reply sql/7_0_0-to-7_0_1_upgrade.sql:198]: problem with setting to `25` would be the possibility of clobbering somebody's already established list

Comment (sjpadgett) [reply sql/7_0_0-to-7_0_1_upgrade.sql:198]: That's why you re-sequence lists in upgrade script using vars. There are several examples in past upgrades. Changing the data type for accommodate one new item is not the answer IMO.

Comment (stephenwaite) [reply sql/7_0_0-to-7_0_1_upgrade.sql:198]: the other string items haven't been working all along as well

Comment (sjpadgett) [reply sql/7_0_0-to-7_0_1_upgrade.sql:198]: Just do a query to get the max id value and add 1 to it then insert new item. If you want I can write one for you.

Comment (tsimonq2) [reply sql/7_0_0-to-7_0_1_upgrade.sql:198]: Why make it more complicated than it has to be? I can see the alternative of simply ensuring all those values are converted over to integers, but why make more database modifications than necessary when the codebase changes are minimal?

Comment (stephenwaite) [reply sql/7_0_0-to-7_0_1_upgrade.sql:198]: @sjpadgett's right about the implications of other folks using the prescription's table

and the sql max id thing is pretty cool in the upgrade script

Comment (tsimonq2) [reply sql/7_0_0-to-7_0_1_upgrade.sql:198]: Let me mull over a better approach over lunch and get back to you - either way I didn't consider that case fully, and I'm starting to think you're correct. If the PR looks drastically different I'll close this one and open a new one.

**Comment by sjpadgett** [comment]
Hey guys here is an approach that should simplify. If there are others added previously then need to delete and reinsert.

```
#IfNotRow2D list_options list_id drug_route title By Mouth
Set @max = (SELECT MAX(Convert(option_id, Signed Integer)) FROM list_options WHERE list_id = 'drug_route');
INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`, `notes`, `codes`) VALUES ('drug_route', @max+1, 'By Mouth', 1, 0, 'PO', 'NCI-CONCEPT-ID:C38288');
#EndIf
```

**Comment by tsimonq2** [comment]
Hey @sjpadgett, thanks for that, it'll make things a little bit easier. I'll just force-push to this PR since our discussion is already here.

I think we really also need to convert over the existing string values, I'll have something here before the end of the day.

**Comment by tsimonq2** [comment]
@sjpadgett @stephenwaite The solution I force-pushed just now seems like it would be the best of both worlds. Let me know what you think. Thanks :)

File: sql/7_0_0-to-7_0_1_upgrade.sql
Comment (sjpadgett) [inline sql/7_0_0-to-7_0_1_upgrade.sql:224]: These will duplicate. One is looking at option_id bymouth while the last is looking for title By Mouth.
Maybe delete existing instead of updating

Comment (tsimonq2) [reply sql/7_0_0-to-7_0_1_upgrade.sql]: @sjpadgett How does it look now?

Comment (sjpadgett) [reply sql/7_0_0-to-7_0_1_upgrade.sql]: The only thing I don't understand is when did the the items with string values get put in? My list doesn't have them on my master dev version.

Comment (stephenwaite) [reply sql/7_0_0-to-7_0_1_upgrade.sql]: `4.2.2 to 5.0.0`

File: sql/7_0_0-to-7_0_1_upgrade.sql
Comment (sjpadgett) [inline sql/7_0_0-to-7_0_1_upgrade.sql:218]: You don't need this for a delete. It won't hurt anything though to leave it!

**Review by sjpadgett — APPROVED** [review]
Looks good to me. Will wait until if Brady has any comments else I'll merge in the morning.
Thanks Simon for the fix.

**Comment by bradymiller** [comment]
@sjpadgett @stephenwaite @tsimonq2
hi all, very busy PR :) Note drug_route is also used in immunizations, so this will break all those mappings. Makes much more sense to allow strings in drug_route. Then won't break all those mappings used in immunization and any other custom code/forms that are using the drug_route list. Would allowing strings in prescription route break anything?

**Comment by sjpadgett** [comment]
> Would allowing strings in prescription route break anything?

It was broken when strings were allowed back in v5.0.0. You're right though this will break Immunizations so sorry.
However you look at it some legacy data is broken. It will be even more broke doing my suggestion so I suppose it will need to be fixed in prescriptions.

I'll fix and put up a PR due to the fact I mislead. Thanks Brady.

**Comment by bradymiller** [comment]
Sometimes it boggles my mind how long a bug like this can exist (ie. any of those string based routes have been breaking in prescription since 5.0.0 :) )

**Comment by sjpadgett** [comment]
see https://github.com/openemr/openemr/pull/5756

**Pull request closed without merging**
